From e60cd32c681302ddbe1e9307832756de61a87161 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 29 Sep 2021 07:56:12 +0300
Subject: [PATCH] [SW-57369] gaudi2: arc: perform checks using sram full addr

Currently, it is assumed that SRAM address received from FIT file
for uboot load is of ARM viewpoint. This patch add changes to convert
full address received from FIT to ARM viewpoint.

Change-Id: I76aa7d1d43bfc685fd19cccb5604354f08fad953
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 3ecb0ac..9623409 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -37,6 +37,7 @@
 #include <arcboot_export.h>
 #include <pci.h>
 #include <hl.h>
+#include <address_space.h>
 
 #define PRODUCT_ID_STR_SIZE	0x8
 
@@ -147,6 +148,12 @@
 	u32 lsb, msb;
 
 	if (startaddr) {
+		/*
+		 * convert startaddr from full address to ARM viewpoint
+		 * address as the reset address registers of ARM needs to
+		 * be programmed with ARM viewpoint address.
+		 */
+		startaddr =  (startaddr & ARM_VP_ADDR_MASK) | MEM_REGION_MASK;
 		lsb = (u32)startaddr;
 		msb = (u32)(startaddr >> 32);
 		/* Set CPU0 Reset address to U-Boot */
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 4d5c578..58bf6a3 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -28,6 +28,7 @@
  * located on HBM0 offset 0x0
  */
 #define ARM_SDRAM_SIZE		0x300000ull
+#define MEM_REGION_MASK		0x8000000000
 #else
 /*
  * The SRAM region in ARC address extension is mapped to 0x10000000, but the
@@ -39,8 +40,10 @@
 #define SDRAM_FULL_BASE		0x1000FFFFFD000000ull
 #define ARM_ADDR_MASK		0x1FFFFFFull
 #define ARM_SDRAM_SIZE		0x3000000
+#define MEM_REGION_MASK		0x4000000000
 #endif
 
+#define ARM_VP_ADDR_MASK	0x3FFFFFFFFF
 //#define GENERAL_CFG_MSB_ADDR	0x1000007F00000000ull
 #define ARC0_ICCM_FULL_BASE	0x1000007FF7000000ull
 #define ARC0_ICCM_SIZE		0x20000			/* 128KB */
